### ISSUE TITLE
Bug in \Project Files\DLLSources\CvCity.cpp, line 11748: further declaration of "mod" can cause OOS failures

### ISSUE THREAD

**Issue by marvol77** [We-the-People-civ4col-mod/Mod#372]
In line 11'748 of **CvCity.cpp** 
     
`int mod = mod_serv; // Servants are supposed to have a higher chance`

should be replaced by
     
`mod = mod_serv; // Servants are supposed to have a higher chance`

i.e., the additional declaration of `mod` renews its scope (`int` should be removed). Otherwise this is a potential cause for an OOS failure due to uninitialized memory used in line 11'754 for the declaration of "calculatedChance".

**Comment by Nightinggale** [comment]
I'm not following you. The line numbers doesn't match what you mention (which branch are you on?) and the only code I can find matching what your code is this:

```cpp
//default case is servant
int mod = mod_serv;
```

It's mentioned 3 times (two in city, one in player). All 3 are different functions meaning they will not hide declarations from each other. Also I can't find that specific comment at all.

**Comment by marvol77** [comment]
Correct, the method `LbD_try_get_free` was already reworked and fixed in the master branch 👍  (I just checked the "original sources" from WTP 2.8. without comparing with the newest master branch).

**Comment by marvol77** [comment]
This bug still occurs in the "Happiness" branch that the last release (2.8, https://github.com/We-the-People-civ4col-mod/Mod/releases) is based on. However, though fixing it locally, OOS issues are frequently occuring with the happiness branch (btw with the current master branch we hadn't any OOS up to now).

**Comment by raystuttgart** [comment]
@marvol77 
Thanks for reporting!

So just to clarify:

Is this OOS issue really caused by **that line of code**? (CvCity.cpp Line 11748)
(At least that is in my local code version - which might be outdated though.)
```
int mod = mod_serv; // Servants are supposed to have a higher chance
```

And yes, you are **completely correct**, it should be coded like this:
(Although I can hardly imagine that it could cause desyncs - but I might be wrong.)
```
mod = mod_serv; // Servants are supposed to have a higher chance
```
---------

Or is this simply about OOS errors still occuring.
(Which is something that we know.)

@Nightinggale 

Basically there is nothing to loose to correct the code that way. 
(Unless it already is corrected in GIT of course.)

If this really fixes the OOS that would be awesome !!!!

**Comment by raystuttgart** [comment]
@Nightinggale , @marvol77 

I compiled a **new DLL** with the fix and also **uploaded** to the **CivFanatics forum**.
I have **asked for help** from community (MP players) to **verify** if it really fixes the **OOS desyncs**.

I also did commit and push to **Branch "Happiness"**. 
(Since then there were no other code changes to my knowledge.)

**Comment by marvol77** [comment]
@raystuttgart , thank you very much for your effort. We like your mod very, very much and we are pleased to contribute some little help. 

> (Although I can hardly imagine that it could cause desyncs - but I might be wrong.)

If `mod` is uninitialized, then `if (randomValue > calculatedChance)`, l. 11761, can differ between players, which again only creates/changes a unit for some of all players in line 11768 (though more an assumption than a proof).
Btw, line 11748 was definitely reached in one of hour games, because it caused an exception in debug mode :) 

> If this should really prove to fix the OOS desyncs

Unfortunatly, it did **not** fix OOS issues for us at all. I fixed it locally, rebuilt, and we played again. Honestly, this corresponds to the release version and does not include the latest changes of the happiness branch. We'll check within the next days with the current state of the branch. 

> Since then there were no other code changes to my knowledge.

At least some files of the Assets folder changed.

Just another remark/question without deeper check: is `remove` from line 11770
`bool remove = removePopulationUnit(convUnit, false, (ProfessionTypes) GC.getCivilizationInfo(GET_PLAYER(getOwnerINLINE()).getCivilizationType()).getDefaultProfession());` 
intended to be used for `bKill` in line 11771
`GeneratedUnit->convert(convUnit, true);`
(instead of `true`)?

**Comment by raystuttgart** [comment]
The "Release Version" was created from the Branch "Happiness", which was the most current branch.
Since the Release I have not changed any code in "Branch Happiness" until the fix yesterday.

So if you use the code from Release Version, it should be the completely same.
(Except the small fix for "mod = mod_serv" mentioned in this thread.)

The only real difference between "Release" and "Happiness" should be in files like Changelog, Credits and other text files.

**Comment by raystuttgart** [comment]
@marvol77 

You seem to be a quite skilled programmer-

Are you also around in our forum at CivFanatics?
Would you be interested to join us?

If yes, please write me in a a private conversation in the forum at CivFanatics or simply leave a post in the forum.

**Comment by raystuttgart** [comment]
I think that was already fixed now.
